tracing/traced_file.cc

```diff
@@ -0,0 +1,219 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <seastar/core/future.hh>
+#include <seastar/core/print.hh>
+
+#include "tracing/traced_file.hh"
+
+using namespace seastar;
+
+class traced_file_impl : public file_impl {
+private:
+    file _f;
+
+    tracing::trace_state_ptr _trace_state;
+    sstring _trace_prefix;
+
+public:
+    unsigned _memory_dma_alignment;
+    unsigned _disk_read_dma_alignment;
+    unsigned _disk_write_dma_alignment;
+
+    traced_file_impl(file, tracing::trace_state_ptr, sstring);
+
+    virtual future<size_t> write_dma(uint64_t pos, const void* buf, size_t len, const io_priority_class& pc) override;
+    virtual future<size_t> write_dma(uint64_t pos, std::vector<iovec> iov, const io_priority_class& pc) override;
+    virtual future<size_t> read_dma(uint64_t pos, void* buf, size_t len, const io_priority_class& pc) override;
+    virtual future<size_t> read_dma(uint64_t pos, std::vector<iovec> iov, const io_priority_class& pc) override;
+    virtual future<> flush(void) override;
+    virtual future<struct stat> stat(void) override;
+    virtual future<> truncate(uint64_t length) override;
+    virtual future<> discard(uint64_t offset, uint64_t length) override;
+    virtual future<> allocate(uint64_t position, uint64_t length) override;
+    virtual future<uint64_t> size(void) override;
+    virtual future<> close() override;
+    virtual std::unique_ptr<seastar::file_handle_impl> dup() override;
+    virtual subscription<directory_entry> list_directory(std::function<future<> (directory_entry de)> next) override;
+    virtual future<temporary_buffer<uint8_t>> dma_read_bulk(uint64_t offset, size_t range_size, const io_priority_class& pc) override;
+};
+
+traced_file_impl::traced_file_impl(file f, tracing::trace_state_ptr trace_state, sstring trace_prefix)
+    : _f(std::move(f)), _trace_state(std::move(trace_state)), _trace_prefix(std::move(trace_prefix)),
+      _memory_dma_alignment(_f.memory_dma_alignment()), _disk_read_dma_alignment(_f.disk_read_dma_alignment()),
+      _disk_write_dma_alignment(_f.disk_write_dma_alignment()) {}
+
+future<size_t>
+traced_file_impl::write_dma(uint64_t pos, const void* buf, size_t len, const io_priority_class& pc) {
+    tracing::trace(_trace_state, "{} scheduling DMA write of {} bytes at position {}", _trace_prefix, len, pos);
+    return get_file_impl(_f)->write_dma(pos, buf, len, pc).then([this, pos, len] (size_t ret) {
+        tracing::trace(_trace_state, "{} finished DMA write of {} bytes at position {}, successfully wrote {} bytes", _trace_prefix, len, pos, ret);
+        return ret;
+    }).handle_exception([this, pos, len] (std::exception_ptr e) {
+        tracing::trace(_trace_state, "{} failed DMA write of {} bytes at position {}, exception: {}", _trace_prefix, len, pos, e);
+        return make_exception_future<size_t>(std::move(e));
+    });
```

Comment:
This generates two continuations, since write_dma() always defers. Better to use then_wrapped() and internally differentiate between the two cases.

Reply:
done